user_tools/src/as_pytools/storagelib/aspath.py

```diff
@@ -0,0 +1,329 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Abstract representation of a file path that can access local/URI values.
+Similar to cloudpathlib project, this implementation uses dict registry to
+register an implementation. However, the path representation is built on top of
+pyArrow FS API. As a result, there is no need to write a full storage client to
+access remote files. This comes with a tradeoff in providing limited set of file
+operations.
+"""
+
+import abc
+import sys
+from collections import defaultdict
+from functools import cached_property
+from pathlib import Path as PathlibPath
+from typing import Union, Type, TypeVar, Any, Dict, Callable, overload, Optional, TYPE_CHECKING, List
+
+from pyarrow import fs
+from pyarrow.fs import FileType, FileSystem, FileInfo
+from pydantic import ValidationError, model_validator, FilePath, AnyHttpUrl, StringConstraints, TypeAdapter
+from pydantic.dataclasses import dataclass
+from pydantic_core import PydanticCustomError
+from typing_extensions import Annotated
+
+from ..exceptions import (
+    InvalidProtocolPrefixError,
+    FSMismatchError, ASFileExistsError
+)
+
+from ..utils.util import get_path_as_uri
+
+if sys.version_info >= (3, 10):
+    from typing import TypeGuard
+else:
+    from typing_extensions import TypeGuard
+if sys.version_info >= (3, 11):
+    from typing import Self
+else:
+    from typing_extensions import Self
+
+if TYPE_CHECKING:
+    from .fs import AsFs
+
+
+CSPPathString = Annotated[str, StringConstraints(pattern=r'^\w+://.*')]
+
+
+class ASPathImplementation:
```

Comment:
nit: could be more consistent with AS/As prefix in `aspath.py` and `fs.py`

Reply:
Done